File: backend/app/routers/network.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.auth import require_admin
from app.crypto import decrypt, encrypt
from app.database import get_db
from app.models import PC, NetworkInfo
from app.schemas import NetworkIn, NetworkOut

router = APIRouter(prefix="/pcs", tags=["network"])


def _get_pc(db: Session, pc_id: str) -> PC:
    pc = db.get(PC, pc_id)
    if pc is None:
        raise HTTPException(status_code=404, detail="PC not found")
    return pc


@router.get("/{pc_id}/network", response_model=NetworkOut)
def get_network(pc_id: str, db: Session = Depends(get_db)):
    _get_pc(db, pc_id)
    info = db.execute(
        select(NetworkInfo).where(NetworkInfo.pc_id == pc_id)
    ).scalar_one_or_none()
    if info is None:
        return NetworkOut(pc_id=pc_id, ip_address=None, mac_address=None, notes=None)
    return NetworkOut(
        pc_id=pc_id,
        ip_address=decrypt(info.ip_address),
        mac_address=decrypt(info.mac_address),
        notes=info.notes,
    )


@router.put("/{pc_id}/network", response_model=NetworkOut, dependencies=[Depends(require_admin)])
def put_network(pc_id: str, payload: NetworkIn, db: Session = Depends(get_db)):
    _get_pc(db, pc_id)
    info = db.execute(
        select(NetworkInfo).where(NetworkInfo.pc_id == pc_id)
    ).scalar_one_or_none()
    if info is None:
        info = NetworkInfo(pc_id=pc_id)
        db.add(info)
    info.ip_address = encrypt(payload.ip_address)
    info.mac_address = encrypt(payload.mac_address)
    info.notes = payload.notes
    db.commit()
    return NetworkOut(
        pc_id=pc_id,
        ip_address=payload.ip_address,
        mac_address=payload.mac_address,
        notes=payload.notes,
    )

File: backend/routes/network.js
import { Router } from 'express';
import { PC, NetworkInfo } from '../config/models.js';
import { encrypt, decrypt } from '../utils/crypto.js';
import { requireAdmin } from '../middleware/auth.js';

const router = Router();

router.get('/:pcId/network', async (req, res, next) => {
  try {
    const pc = await PC.findById(req.params.pcId).lean();
    if (!pc) return res.status(404).json({ detail: 'PC not found' });
    const info = await NetworkInfo.findOne({ pc_id: req.params.pcId }).lean();
    if (!info) return res.json({ pc_id: req.params.pcId, ip_address: null, mac_address: null, notes: null });
    res.json({ pc_id: req.params.pcId, ip_address: decrypt(info.ip_address), mac_address: decrypt(info.mac_address), notes: info.notes });
  } catch (err) { next(err); }
});

router.put('/:pcId/network', requireAdmin, async (req, res, next) => {
  try {
    const pc = await PC.findById(req.params.pcId).lean();
    if (!pc) return res.status(404).json({ detail: 'PC not found' });
    const { ip_address, mac_address, notes } = req.body;
    let info = await NetworkInfo.findOne({ pc_id: req.params.pcId }).lean();
    const data = { pc_id: req.params.pcId, ip_address: encrypt(ip_address), mac_address: encrypt(mac_address), notes: notes || null };
    if (info) {
      await NetworkInfo.updateOne({ pc_id: req.params.pcId }, { $set: data });
    } else {
      await NetworkInfo.create(data);
    }
    res.json({ pc_id: req.params.pcId, ip_address, mac_address, notes: notes || null });
  } catch (err) { next(err); }
});

export default router;
